File: backend/app/api/report.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies.auth import get_current_user

from app.repositories.interview_repository import (
    get_interview_by_id,
)

from app.repositories.report_repository import (
    get_reports_by_user,
)

from app.services.report_service import (
    generate_interview_report,
)


# --------------------------------------------------
# Reports History Router
# Endpoint: GET /reports
# --------------------------------------------------

reports_router = APIRouter(
    prefix="/reports",
    tags=["Reports"],
)


@reports_router.get("")
def get_all_reports(
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """
    Return all interview reports belonging
    to the currently authenticated user.
    """

    reports = get_reports_by_user(
        db,
        current_user.id,
    )

    result = []

    for report in reports:

        interview = report.interview

        result.append(
            {
                "id": report.id,
                "interview_id": report.interview_id,

                "company": interview.company,
                "role": interview.role,

                "overall_score": report.overall_score,
                "technical_score": report.technical_score,
                "communication_score": report.communication_score,
                "confidence_score": report.confidence_score,

                "summary": report.summary,
                "strengths": report.strengths,
                "weaknesses": report.weaknesses,
                "recommendations": report.recommendations,
                "learning_roadmap": report.learning_roadmap,
                "hiring_recommendation": report.hiring_recommendation,

                "created_at": report.created_at,
            }
        )

    return result


# --------------------------------------------------
# Single Interview Report Router
# Endpoint: GET /interviews/{interview_id}/report
# --------------------------------------------------

router = APIRouter(
    prefix="/interviews",
    tags=["Reports"],
)


@router.get("/{interview_id}/report")
def get_report(
    interview_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):

    interview = get_interview_by_id(
        db,
        interview_id,
    )

    if interview is None:
        raise HTTPException(
            status_code=404,
            detail="Interview not found",
        )

    if interview.user_id != current_user.id:
        raise HTTPException(
            status_code=403,
            detail="Access denied",
        )

    return generate_interview_report(
        db,
        interview_id,
    )

File: frontend/src/api/report.ts
import api from "./axios";

/* ---------- Types ---------- */

export interface Report {
  id?: number;
  interview_id: number;

  overall_score: number;
  technical_score: number;
  communication_score: number;
  confidence_score: number;

  summary: string;

  strengths: string | string[];
  weaknesses: string | string[];
  recommendations: string | string[];
  learning_roadmap: string | string[];

  hiring_recommendation: string;

  created_at?: string;
}

/* ---------- Get Single Interview Report ---------- */

export async function getReport(
  interviewId: number | string
): Promise<Report> {
  const { data } = await api.get<Report>(
    `/interviews/${interviewId}/report`
  );

  return data;
}

/* ---------- Get All Reports ---------- */

/*
  Your backend currently does NOT have GET /reports.

  This function is kept for future use when we add
  the reports history endpoint.
*/

export async function getReports(): Promise<Report[]> {
  try {
    const { data } = await api.get<Report[]>("/reports");

    return data;
  } catch (error: any) {
    if (error.response?.status === 404) {
      console.warn(
        "Reports history endpoint is not available yet."
      );

      return [];
    }

    throw error;
  }
}